docs/sphinx-doc/source/environment_variables.rst
*********************************************
Environment Variables
*********************************************

These environment variables can be used to control OpenCL behavior and provide visibility for debugging.

.. envvar:: TI_OCL_KEEP_FILES

    When OpenCL C kernels are compiled for DSPs, the result is a binary
    .out file in the /tmp sub-directory. They are then subsequently available
    for download to the DSPs for running. The process of compiling generates
    several intermediate files for each source file. OpenCL typically removes
    these temporary files. However, it can sometimes be useful to inspect these
    files.  This environment variable can be set to instruct the runtime to
    leave the temporary files in /tmp. Inspecting the assembly file associated
    with the out file, can be useful to see how well your code was optimized.

.. envvar:: TI_OCL_DEBUG

    Setting this environment variable modifies the execution of OpenCL
    applications to enable debug of the OpenCL C kernels.  If your application
    uses on-line compilation (i.e. it compiles from a string rather than a
    binary), then that on-line compilation asserts the debug flag to the
    compiler. (If the application is using off-line compilation, i.e. creating
    a program from binary, then you would need to pass -g as an option to the
    off-line compiler clocl. The OpenCL runtime pauses your application before
    dispatch of all kernels. While paused the runtime indicates to the user
    that a kernel dispatch is pending.

    Currently two debugging methods are supported by setting TI_OCL_DEBUG to
    either "gdb" or "ccs".  If set to "gdb", the runtime provides the gdbc6x
    commands to connect to the DSP and setup appropriate breakpoints.  If set
    to "ccs", the runtime provides the Code Composer Studio (CCS) instructions
    to connect to the DSP and setup appropriate breakpoints.  With either
    method, the runtime forces all kernels and work-groups within kernels
    to execute only on DSP core 0.  Details can be found in
    :doc:`debug/index`.

.. envvar:: TI_OCL_CACHE_KERNELS

    On-line compilation of kernels is a useful feature for portable OpenCL
    programs. All the detail required to compile kernels for devices is
    encapsulated in the OpenCL API calls. However, on-line compilation for the
    DSPs can be time-consuming. Setting this environment variable causes the
    OpenCL runtime to perform one on-line compilation of your kernels and cache
    the results in a database in /tmp. Running the application again without
    modification of the kernel source or the options used to compile it,
    results in the compilation step being bypassed, and the use of the cached
    kernel binary.

    .. Warning::

        If OpenCL C kernels call standard C code, modifications to the standard
        C code are not seen by the OpenCL runtime and a cached result may be
        used when it is not appropriate. If calling standard C code, either
        disable this environment variable or clean the cache anytime the
        standard C code is modified.


    .. Warning::

        Using this environment variable causes persistent data to accumulate in
        /tmp, and /tmp may grow to capacity causing run-time errors. If this
        occurs, remove cached objects in /tmp or increase the size of the /tmp
        partition. To explicitly remove the cache, execute the
        command: :command:`rm -f /tmp/opencl*`.

    .. Note::

        The OpenCL compilation cache is automatically removed during a Linux reboot

.. envvar::  TI_OCL_COMPUTE_UNIT_LIST

    Specify the compute units available to the OpenCL runtime as a comma
    separated list of compute unit indices starting at 0.  The specified
    compute unit list must be consecutive, e.g. ``"1,2,3,4"`` on K2H.
    If the environment variable is not specified, the runtime defaults to
    using all the compute units available on the device (or starting from
    OpenCL product version 1.1.13, all available compute units specified in
    ``/etc/ti-mctd/ti_mctd_config.json``).

    Example usage on AM572x:

    .. code-block:: bash
        :caption: runs the vecadd kernel only on DSP1

        -> TI_OCL_COMPUTE_UNIT_LIST="0" ./vecadd

    .. code-block:: bash
        :caption: runs the vecadd kernel only on DSP2

        -> TI_OCL_COMPUTE_UNIT_LIST="1" ./vecadd


    .. code-block:: bash
        :caption: runs the vecadd kernel on both DSP1 and DSP2 (default behavior)

        -> TI_OCL_COMPUTE_UNIT_LIST="0, 1" ./vecadd


    .. Warning::

        Prior to OpenCL product version 1.1.13, this environment variable is
        available only on AM572x.

.. envvar::  TI_OCL_LOAD_KERNELS_ONCHIP

    By default, OpenCL kernel related code and global data is allocated out of
    DDR memory. If this environment variable is set, kernel related code and
    global data is allocated out of MSMC memory.

    ..Warning::

        Rarely used K2x only feature, will be deprecated starting with OpenCL version 1.1.13.0.


.. envvar::  TI_OCL_CPU_DEVICE_ENABLE

    Currently, OpenCL ARM CPU devices only support native kernels (see the
    OpenCL 1.1 spec for a description of native kernels). As a result, the ARM
    CPU is not, by default, treated as a COMPUTE DEVICE when doing an OpenCL
    platform query. If your application only uses the ARM CPU for native
    kernels, then this environment variable can be used to enable it as a
    COMPUTE DEVICE for OpenCL. Enqueueing NDRangeKernels or Tasks to the CPU is
    not supported, even when this environment variable is set.

.. envvar::  TI_OCL_WORKER_SLEEP

    .. Warning::
        Removed in OpenCL 1.1.17, no longer required.

    The OpenCL runtime starts a new CPU thread for every OpenCL command
    queue defined in your application. These threads
    manage the OpenCL command queues and the communication
    between the CPU and the device to which the command queue is associated. If
    there are any OpenCL kernels actively running on the device, the thread
    assigned to monitor the communication with the device on behalf of those
    kernels consumes CPU resources, checking the status of those kernels.
    This environment variable can be used to provide a level of control on how
    much CPU resource is consumed. When TI_OCL_WORKER_SLEEP is unset, the
    OpenCL runtime uses more CPU capacity to ensure the fastest turnaround
    latency on kernel execution. When the TI_OCL_WORKER_SLEEP environment
    variable is set to a number of microseconds, it degrades the
    turnaround latency for a kernel execution to reduce the CPU capacity
    needed to monitor the kernel. If an application is not performance limited
    by CPU cycles or if the application enqueues many fine-grained kernels,
    then having the TI_OCL_WORKER_SLEEP environment variable unset is
    appropriate. In the opposite cases, when CPU cycles are limiting the
    performance of an application or if fewer, but longer running kernels are
    enqueued, then setting TI_OCL_WORKER_SLEEP to some number of microseconds
    is appropriate. The correct number of microseconds to use depends on
    the execution platform and the particular application. However, using a
    microseconds value in the range from 80 to 150 is a reasonable starting
    point.

.. envvar::  TI_OCL_ENABLE_FP64

    The C66x DSP is double precision floating point capable and all the optional
    features in the OpenCL specification for double precision floating point
    are supported in this OpenCL implementation, except for the requirement
    that double FP support include subnormal behavior or graceful underflow.
    The 64-bit floating point hardware on the C66x DSP does not support
    subnormal behavior. It supports flush to zero behavior. To support
    subnormal behavior for doubles would require software emulation that would
    entail a significant performance penalty versus the hardware capabilities
    of the C66x DSP. Therefore, by default the platform and devices supported in
    the TI OpenCL implementation do not report support for double floating
    point. That is, if the platform or device is queried for extensions,
    cl_khr_fp64 is not listed by default. Additionally the OpenCL C predefined
    macro cl_khr_fp64 is not be defined by default. When the
    TI_OCL_ENABLE_FP64 environment variable is set, the TI OpenCL
    implementation reports support for double floating point, i.e.
    cl_khr_fp64 is listed as an extension for the platform and the DSP
    device and cl_khr_fp64 is defined when compiling OpenCL C kernels.
    This environment variable controls whether the OpenCL implementation
    reports support for double. However, double, all double vector types and
    all built-in functions using doubles are supported and available without
    regard to the setting of this environment variable.

.. envvar::  TI_OCL_VERBOSE_ERROR

    The OpenCL specification provides a well-defined mechanism for returning
    error codes from API functions. However, It is often the case that a
    generic error code is returned for differing reasons. When this
    environment variable is set, the OpenCL runtime may print more description
    error messages in addition to the defined return code error mechanism.

.. envvar::  TI_OCL_WG_SIZE_LIMIT

    OpenCL provides a query to a device for the maximum number of work-items
    allowed in a work-group. The DSP device in TI's implementation allows a
    large number of work-items per work-group. Other OpenCL implementations have
    much smaller max work-group size limit. When running code designed and
    optimized for other OpenCL implementations, this environment variable can
    be used to limit the max work-group size reported.

.. envvar::  TI_OCL_PROFILING_EVENT_TYPE

    Specifies the hardware event type to profile. The two basic divisions,
    stall cycle events and memory events, are described in :doc:`profiling`.
    If 1 is specified, OpenCL runtime will profile a stall cycle event.
    If 2 is specified, OpenCL runtime will profile one or two memory event(s).
    Otherwise, profiling is disabled.

.. envvar::  TI_OCL_PROFILING_EVENT_NUMBER1

    Specifies the event number to profile. The exact value of this variable
    represents the offset from either AET_GEM_STALL_EVT_START or
    AET_GEM_MEM_EVT_START, depending on the event type.
    For a full event list, see :doc:`profiling`.

.. envvar::  TI_OCL_PROFILING_EVENT_NUMBER2

    Specifies the second memory event number to profile.  Can be skipped.

.. envvar::  TI_OCL_PROFILING_STALL_CYCLE_THRESHOLD

    Specified the threshold of stall cycles to count.  Only stall events with
    stall cycles higher than this threshold are captured in the counter.
    Default value in OpenCL runtime is 1, i.e. all stall events are captured.



debian/changelog
ti-opencl (01.01.08.00-0ubuntu0~ppa1) trusty; urgency=medium

  * Added K2G platform support 

 -- TI Keystone PPA <ti-keystone-ppa@list.ti.com>  Fri, 18 Mar 2016 13:29:07 -0500

ti-opencl (01.01.06.00-0ubuntu0~ppa1) trusty; urgency=medium

  * Added L1D scratchpad management functions
  * Updated documentation

 -- TI Keystone PPA <ti-keystone-ppa@list.ti.com>  Thu, 10 Sep 2015 16:00:09 -0500

ti-opencl (01.01.05.00-0ubuntu0~ppa1) trusty; urgency=medium

  * Defects fixed:
    - Update KernelEvent to properly ref_counts memory objects
    - Perform cache wbInv before updating MAR registers
    - Send response to host on EXIT to ensure DSP exit is complete
  * Updated documentation
  * Fixed sgemm example

 -- TI Keystone PPA <ti-keystone-ppa@list.ti.com>  Fri, 21 Aug 2015 15:31:21 -0500

ti-opencl (01.01.04.00-0ubuntu0~ppa1) UNRELEASED; urgency=medium

  * Added an example, float_compute
  * Removed busy waiting in the host runtime and fixed a race condition
    related to mailbox communication 

 -- Ajay Jayaraj <ajayj@uda0323430>  Mon, 15 Jun 2015 11:18:04 -0500

ti-opencl (01.01.02.00-0ubuntu0~ppa1) trusty; urgency=medium

  * AM57x : enabled dispatch of out of order tasks to both DSP cores

 -- TI Keystone PPA <ti-keystone-ppa@list.ti.com>  Wed, 20 May 2015 10:38:59 -0500

ti-opencl (01.01.01.02-0ubuntu0~ppa1) trusty; urgency=medium

  * Give larger alignment to buffers allocated in extended memory to help MPAX
    allocation.

  * Provide better error reporting when error occurs in OpenCL runtime.

 -- TI Keystone PPA <ti-keystone-ppa@list.ti.com>  Tue, 12 May 2015 15:40:53 -0500

ti-opencl (01.01.01.01-0ubuntu0~ppa1) trusty; urgency=medium

  * Make __malloc/free_ddr/msmc() thread safe, use Lock to guard the
    bookkeeping map accesses.

  * Update to always flush buffers after kernel execution, even when kernel
    has only WRITE_ONLY buffers

  * Update MPX examples to query available memory and allocate buffers
    accordingly

  * Add OSAL critical section support in RM client code that allocates QMSS
    resources.

 -- TI Keystone PPA <ti-keystone-ppa@list.ti.com>  Fri, 17 Apr 2015 15:48:32 -0500

ti-opencl (01.01.00.03-0ubuntu0~ppa1) trusty; urgency=medium

  * Added signal handler in opencl to cleanup if user hits ctrl-c during
    OpenCL execution.

  * Updated ARM makefile install to include new monitor file 'dsp_syms.obj'.

 -- TI Keystone PPA <ti-keystone-ppa@list.ti.com>  Tue, 14 Apr 2015 15:03:01 -0500

ti-opencl (01.01.00.02-0ubuntu0~ppa1) trusty; urgency=medium

  * Modified the monitor mechanism for mapping builtin functions names to
    address from being text based to being object based.  This speeds link
    time significantly.

  * Update version scheme for debian packaging to ww.xx.yy.zz format.
    Where:
        ww.xx - Maps to supported OpenCL specification version. Currently 1.1.
        yy    - TI release
        zz    - TI patch release

 -- TI Keystone PPA <ti-keystone-ppa@list.ti.com>  Fri, 10 Apr 2015 15:38:44 -0500

ti-opencl (01.01.00.01-0ubuntu0~ppa1) trusty; urgency=medium

  * Make OpenCL use Resource Manager (RM) to allocate QMSS resources on the
    ARM side and pass to DSP for QMSS use.

 -- TI Keystone PPA <ti-keystone-ppa@list.ti.com>  Mon, 23 Mar 2015 16:04:26 -0500

ti-opencl (01.01.00.00-0ubuntu0~ppa3) trusty; urgency=medium

  * Fixed a C++ global objects destruction ordering problem that can cause
    segmentation fault at OpenCL application exit.

 -- TI Keystone PPA <ti-keystone-ppa@list.ti.com>  Thu, 05 Mar 2015 09:38:59 -0600

ti-opencl (01.01.00.00-0ubuntu0~ppa2) trusty; urgency=medium

  * Change use of QMSS resources to avoid conflicts with mpi.
  * Fix clocl issue where incorrect path was specified for the intermediate
    .bc file when clocl invoked the C6x compiler.
  * Added the -a/--alias option to clocl. This option must be enabled when the
    user creates overlapping buffers via USE_HOST_PTR on pointers pointing to
    an object allocated via __malloc_ddr or __malloc_msmc.

 -- TI Keystone PPA <ti-keystone-ppa@list.ti.com>  Wed, 04 Mar 2015 14:17:42 -0600

ti-opencl (01.01.00.00-0ubuntu0~ppa1) trusty; urgency=medium

  * Update version scheme for debian packaging to ww.xx.yy.zz format.
    Where:
        ww.xx - Maps to supported OpenCL specification version. Currently 1.1.
        yy    - TI release
        zz    - TI patch release

 -- TI Keystone PPA <ti-keystone-ppa@list.ti.com>  Sun, 15 Feb 2015 13:21:14 -0600

ti-opencl (1.0.0-0ubuntu0~ppa3) trusty; urgency=medium

  * Bump version to ppa3 because of failed build.

 -- TI Keystone PPA <ti-keystone-ppa@list.ti.com>  Fri, 30 Jan 2015 14:12:42 -0600

ti-opencl (1.0.0-0ubuntu0~ppa2) trusty; urgency=medium

  * Change 'matmpy' OpenCL examples to use less MSMC memory.

 -- TI Keystone PPA <ti-keystone-ppa@list.ti.com>  Thu, 29 Jan 2015 15:49:48 -0600

ti-opencl (1.0.0-0ubuntu0~ppa1) trusty; urgency=medium

  * Faster workgroup handling on the dsp
  * The file /var/lock/opencl will now be locked just in time within a process
    at the point when OpenCL is first used. Previously, the lock was created at
    startup of the process, even if OpenCL was not invoked for some time.
  * The OpenCL library creates a host cpu thread to monitor traffic to and from
    the DSP device. Previously that thread would consume a host CPU. It has
    been improved to not consume any CPU resources unless there is an OpenCL
    active kernel on the DSP device. When there is an active kernel on the DSP
    device, it will consume resources at a nice level of 4. This can reduce the
    CPU impact of the monitoring thread. Additionally an environment variable
    TI_OCL_WORKER_SLEEP can be set to a non-negative number that represents the
    number of micro seconds that the monitoring thread should sleep after each
    device check. This will reduce the cpu overhead of the monitor thread but
    will increase the round trip time for any given kernel by the sleep value

host/clocl/options.cpp
#include <stdio.h>     /* for printf */
#include <stdlib.h>    /* for exit */

#if _MSC_VER
#include "getopt.h"
#define strtok_r strtok_s
#else
#include <getopt.h>
#endif
#include <string.h>

#include <string>
#include <vector>
#include <iostream>
#include <iterator>
#include <algorithm>

#include "file_manip.h"

using std::cout;
using std::endl;
using std::string;
using std::vector;
using std::ostream_iterator;

int opt_help    = 0;
int opt_verbose = 0;
int opt_keep    = 0;
int opt_debug   = 0;
int opt_symbols = 0;
int opt_lib     = 0;
int opt_txt     = 0;
int opt_w       = 0;
int opt_Werror  = 0;
int opt_builtin = 0;
int opt_tmpdir  = 0;
int opt_version = 0;
int opt_alias   = 0;

string cl_options;
string cl_incdef;
string opts_other;
vector<string> files_clc;
vector<string> files_c;
string         files_other;

#define STRINGIZE(x) #x
#define STRINGIZE2(x) STRINGIZE(x)

void print_version()
{
    printf("OpenCL 1.1 TI product version "
           STRINGIZE2(_PRODUCT_VERSION)
           " (" __DATE__ " " __TIME__ ")\n\n");
}

/******************************************************************************
* void print_options()
******************************************************************************/
void print_options()
{
    cout << endl;

    if (opt_keep)    printf ("Option keep   : on\n");
    if (opt_debug)   printf ("Option debug  : on\n");
    if (opt_lib)     printf ("Option lib    : on\n");
    if (opt_txt)     printf ("Option txt    : on\n");
    if (opt_w)       printf ("Option w      : on\n");
    if (opt_Werror)  printf ("Option Werror : on\n");
    if (opt_alias)   printf ("Option alias  : on\n");
    if (opt_symbols) printf ("Option symbols: on\n");
    //if (opt_builtin) printf ("Option builtin: on\n");
    //if (opt_tmpdir)  printf ("Option tmpdir : on\n");

    cout << endl;

    cout << "CL C Options  : " << cl_options   << endl;
    cout << "Incls/Defines : " << cl_incdef    << endl;
    if (!files_clc.empty())
        cout << "CL C file     : " << files_clc[0] << endl;
    cout << "Link Files    : " << files_other  << endl;

    cout << endl;

    cout << "Ignored Opts  : " << opts_other   << endl;
    cout << "Ignored Files : ";
    if (files_clc.size() > 1)
        copy(files_clc.begin()+1, files_clc.end(), ostream_iterator<string>(cout, " "));

    copy(files_c.begin(), files_c.end(), ostream_iterator<string>(cout, " "));
    cout << endl << endl;
}

/******************************************************************************
* void print_help()
******************************************************************************/
void print_help()
{
    cout << endl;
    cout << "Usage: clocl [options] <OpenCL C file> [<link files>]" << endl;
    cout << endl;

    cout << "Options passed to clocl are either options to control" << endl;
    cout << "clocl behavior or they are documented OpenCL 1.1 build" << endl;
    cout << "options." << endl;
    cout << endl;
    cout << "The clocl behavior options are: " << endl;
    cout << "   -h, --help    : Print this help screen" << endl;
    cout << "   -v, --verbose : Print verbose messages" << endl;
    cout << "   -k, --keep    : Do not delete temp compilation files" << endl;
    cout << "   -g, --debug   : Generate debug symbols" << endl;
    cout << "   -t, --txt     : Generate object in header form" << endl;
    cout << "   -l, --lib     : Do not link. Stop after compilation." << endl;
    cout << "   -s, --symbols : Keep Symbols." << endl;
    cout << "   -a, --alias   : Assume kernel buffers alias each other" << endl;
    cout << "   --version     : Print OpenCL product." << endl;
    cout << endl;
    cout << "The OpenCL 1.1 build options. Refer to 1.1 spec for desc:" << endl;
    cout << "   -D<name>" << endl;
    cout << "   -D<name>=<val>" << endl;
    cout << "   -I<dir>" << endl;
    cout << "   -w" << endl;
    cout << "   -Werror" << endl;
    cout << "   -cl-single-precision-constant" << endl;
    cout << "   -cl-denorms-are-zero" << endl;
    cout << "   -cl-opt-disable" << endl;
    cout << "   -cl-mad-enable" << endl;
    cout << "   -cl-no-signed-zeros" << endl;
    cout << "   -cl-unsafe-math-optimizations" << endl;
    cout << "   -cl-finite-math-only" << endl;
    cout << "   -cl-fast-relaxed-math" << endl;
    cout << "   -cl-std=<val>" << endl;
    cout << endl;
    exit(-1);
}

/******************************************************************************
* void process_options(int argc, char **argv)
******************************************************************************/
void process_options(int argc, char **argv)
{
    int c;

    while (1) 
    {
        static struct option long_options[] = {

            /*-----------------------------------------------------------------
            * clocl options
            *----------------------------------------------------------------*/
            {"help",    no_argument,  &opt_help,    'h' },
            {"verbose", no_argument,  &opt_verbose, 'v' },
            {"keep",    no_argument,  &opt_keep,    'k' },
            {"debug",   no_argument,  &opt_debug,   'g' },
            {"symbols", no_argument,  &opt_symbols, 's' },
            {"lib",     no_argument,  &opt_lib,     'l' },
            {"txt",     no_argument,  &opt_txt,     't' },
            {"builtin", no_argument,  &opt_builtin, 'b' },
            {"tmpdir",  no_argument,  &opt_tmpdir,  'd' },
            {"alias",   no_argument,  &opt_alias,   'a' },
            {"version", no_argument,  &opt_version,  1  },

            /*-----------------------------------------------------------------
            * opencl 1.1 options
            *----------------------------------------------------------------*/
            {"Werror",                       no_argument, 0,  0  },
            {"cl-std",                       required_argument, 0,  0  },
            {"cl-single-precision-constant", no_argument, 0, 0 },
            {"cl-denorms-are-zero",          no_argument, 0, 0 },
            {"cl-opt-disable",               no_argument, 0, 0 },
            {"cl-mad-enable",                no_argument, 0, 0 },
            {"cl-no-signed-zeros",           no_argument, 0, 0 },
            {"cl-unsafe-math-optimizations", no_argument, 0, 0 },
            {"cl-finite-math-only",          no_argument, 0, 0 },
            {"cl-fast-relaxed-math",         no_argument, 0, 0 },
            {0,                              0,           0,  0 }
        };

        int this_option_optind = optind ? optind : 1;
        int option_index = 0;

        opterr = 0; // prevent getopt from printing warnings

       c = getopt_long_only(argc, argv, "-tadgwsI:D:", long_options,
                            &option_index);
       if (c == -1) break;

       switch (c) 
       {
           case 0:
               {
                string name(long_options[option_index].name);

                if (name == "help"    || name == "verbose"   ||
                    name == "keep"    || name == "debug"     ||
                    name == "lib"     || name == "txt"       ||
                    name == "builtin" || name == "tmpdir"    ||
                    name == "alias"   || name == "symbols"
                   ) break;

                if (name == "cl-std")
                {
                    cl_options += " -cl-std=";
                    cl_options += optarg;
                    break;
                }

                if (name == "Werror")         opt_Werror = 1; // fall-through
                if (name == "cl-opt-disable") opt_debug  = 1; // fall-through

                cl_options += " ";
                cl_options += argv[this_option_optind];
                break;
               }
                
           case 1: 
               {
                 string fname(argv[this_option_optind]);
                 string ext(fs_ext(fname));

                 if      (ext == ".clc") files_clc.push_back(fname);
                 else if (ext == ".cl")  files_clc.push_back(fname);
                 else if (ext == ".c")   files_c.push_back(fname);
                 else                    { files_other += fname; files_other += " "; }

                 break;
               }

           case 't': opt_txt = 1;    break;
           case 'd': opt_tmpdir = 1; break;
           case 'a': opt_alias = 1;  break;
           case 'g': opt_debug  = 1; break;
           case 's': opt_symbols = 1; break;
           case 'w': opt_w = 1; cl_options += " -w"; break;

           case 'D': 
           case 'I': 
                     cl_incdef += " -";
                     cl_incdef += c;
                     cl_incdef += optarg;
                     break;

           case '?':
                opts_other += " ";
                opts_other += argv[this_option_optind];
                break;

           default:
                opts_other += " -";
                opts_other += c;
                break;
        }
    }

    /*-------------------------------------------------------------------------
    * __TI_OCL: if defined, then it is TI OpenCL implementation
    * __TI_OCL_VERSION: 0xXXYYZZWWU, e.g. version 0.12.0 => 0x00120000U
    *------------------------------------------------------------------------*/
    cl_incdef += " -D__TI_OCL";
    char tiocl_ver[64];
    sprintf(tiocl_ver, STRINGIZE2(_PRODUCT_VERSION));
    int ver[4] = { 0, 0, 0, 0 };
    char *firstcall = tiocl_ver, *token, *saveptr;
    for (int i = 0; i < 4; ++i, firstcall = NULL)
    {
        token = strtok_r(firstcall, ".", &saveptr);
        if (token == NULL)  break;
        ver[i] = atoi(token);
        if (ver[i] > 99) ver[i] = 99;  // report error?
    }
    snprintf(tiocl_ver, sizeof(tiocl_ver),
            " -D__TI_OCL_VERSION=0x%02d%02d%02d%02dU",
            ver[0], ver[1], ver[2], ver[3]);
    cl_incdef += tiocl_ver;
    // Otherwise, clang at -O0 will throw those inline function away
    if (opt_debug)  cl_incdef += " -Dinline= ";

    if (opt_version) print_version();
    if (opt_verbose) print_options();
    if (opt_help)    print_help();

    cl_options += " ";
    cl_options += cl_incdef;
}
